src/plugins/intel_gna/src/gna_device.cpp

```diff
@@ -548,16 +548,27 @@ void GNADeviceHelper::open() {
     deviceOpened = true;
 }
 
-void GNADeviceHelper::close() {
-    auto requestsToClose = unwaitedRequestIds;
-    for (auto requestId : requestsToClose) {
+void GNADeviceHelper::stop() {
+    using UnwReqIt = std::pair<bool, UnwaitedRequestIds::value_type>;
+
+    auto firstId = [this]() {
+        std::unique_lock<std::mutex> lockGnaCalls{acrossPluginsSync};
+        auto it = unwaitedRequestIds.begin();
+        return (it == unwaitedRequestIds.end()) ? UnwReqIt(false, 0) : UnwReqIt(true, *it);
+    };
+
+    UnwReqIt it;
+    while ((it = firstId()).first)
         try {
-            waitForRequest(requestId);
+            waitForRequest(it.second);
         } catch (...) {
-            log::warning() << "Request with Id " << requestId << " was not awaited successfully";
+            log::warning() << "Request with Id " << it.second << " was not awaited successfully";
         }
-    }
-    std::unique_lock<std::mutex> lockGnaCalls{ acrossPluginsSync };
+}
+
+void GNADeviceHelper::close() {
+    stop();
```

Comment:
Curious, whether it is still needed.
IMO high chances that's not.
@mkwapul do you see any scenario when this is needed?
(We can try to answer this question after merging this PR)

Reply:
ok, I'll get back to question above.
thank you for approval.

Reply:
it looks like a workaround, because waitForRequest() should wait for the closing of the request.
What of this request will never closed? How long we will check unwaitedRequestIds?

Reply:
moreover Gna2DeviceClose(nGnaDeviceIndex); still has critical issue:
==1194772==ERROR: AddressSanitizer: new-delete-type-mismatch on 0x610000320040 in thread T0:
  object passed to delete has wrong type:
  size of the allocated type:   192 bytes;
  size of the deallocated type: 1 bytes.
    #0 0x7ffff7692e55 in operator delete[](void*, unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:180

Reply:
> it looks like a workaround, because waitForRequest() should wait for the closing of the request. 
prior the fix, no waitForRequest() was called before memory swipe => crash. please point out.

> What if this request will never closed? How long we will check unwaitedRequestIds?
question unrelated to fix, I mean valid also for code prior the fix. But, it's good one though and (if) there is a need to address it, let's do in on behalf of another pull request.

Reply:
> moreover Gna2DeviceClose(nGnaDeviceIndex); still has critical issue: ==1194772==ERROR: >AddressSanitizer: new-delete-type-mismatch on 0x610000320040 in thread T0: object passed to delete has wrong type: size of the allocated type: 192 bytes; size of the deallocated type: 1 bytes. #0 0x7ffff7692e55 in operator delete[](void*, unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:180

fix aims at race described in description section, esspecialy not address at sanitizer warning appearing in the code. It's important of course, but I wouldn't like to intermix other problems with this one unless proved to be part of.

Reply:
> > it looks like a workaround, because waitForRequest() should wait for the closing of the request.
> 
> prior the fix, no waitForRequest() was called before memory swipe => crash. please point out.
> 
> > What if this request will never closed? How long we will check unwaitedRequestIds?
> 
> question unrelated to fix, I mean valid also for code prior the fix. But, it's good one though and (if) there is a need to address it, let's do in on behalf of another pull request.

So we don't need stop() changes to fix the issue? Then I would prefer to exclude them from PR

Reply:
> > it looks like a workaround, because waitForRequest() should wait for the closing of the request.
> 
> prior the fix, no waitForRequest() was called before memory swipe => crash. please point out.
> 
> > What if this request will never closed? How long we will check unwaitedRequestIds?
> 
> question unrelated to fix, I mean valid also for code prior the fix. But, it's good one though and (if) there is a need to address it, let's do in on behalf of another pull request.

No, previously it was a for() loop, now it is in while loop with possible infinite condition

Reply:
> > > it looks like a workaround, because waitForRequest() should wait for the closing of the request.
> > 
> > 
> > prior the fix, no waitForRequest() was called before memory swipe => crash. please point out.
> > > What if this request will never closed? How long we will check unwaitedRequestIds?
> > 
> > 
> > question unrelated to fix, I mean valid also for code prior the fix. But, it's good one though and (if) there is a need to address it, let's do in on behalf of another pull request.
> 
> So we don't need stop() changes to fix the issue? Then I would prefer to exclude them from PR

we do need. previous implementation were leaving possible jobs running. if we left it here it would result in crash.
it's still better to see the app waiting longer for jobs to finish rather than leaving jobs running + swiping memory that is used by them. If app is closing too long, then it may be the user decision to break it.

we can elaborate on jobs waiting forever (?still very rare comparing to the crash occurrence - it's the reason of P1-stopper escalation), but this IMO requires a bit broader discussion on how to handle it correctly (changes lower layers?)

Also, as an extra point, stop() makes access to unwaitedRequestIds thread safe, which required here.

Reply:
> > > it looks like a workaround, because waitForRequest() should wait for the closing of the request.
> > 
> > 
> > prior the fix, no waitForRequest() was called before memory swipe => crash. please point out.
> > > What if this request will never closed? How long we will check unwaitedRequestIds?
> > 
> > 
> > question unrelated to fix, I mean valid also for code prior the fix. But, it's good one though and (if) there is a need to address it, let's do in on behalf of another pull request.
> 
> No, previously it was a for() loop, now it is in while loop with possible infinite condition

Agreed, but in case of infinite loop we are in such a big trouble, it's valuable to be alerted by not exiting IMO? Also, please see my answer above.

Reply:
> Curious, whether it is still needed. IMO high chances that's not. @mkwapul do you see any scenario when this is needed?

@kbruniec, it's not needed here, ie stop() will just be idle. 
HYPOTHETIC scenario: GNADeviceHelper is NOT managed by GNAPlugin and does not play it's second nature role of memory allocator (IMO: cause of all the 'evil' here).
All unwaited tasks are synchronized here then. It's at least (and at most?) nice to have such sync point?

.....GNADeviceHelper has public constructor so (again hypothetically) it's possible for it to exist outside of GNAPlugin.

Reply:
> Agreed, but in case of infinite loop we are in such a big trouble, it's valuable to be alerted by not exiting IMO? Also, please see my answer above.

.... we might use another alert in this case, just exit the application in case some thread cannot be stopped after a few trials
??

plain return from function as it was previously hides problems away

Reply:
> .... we might use another alert in this case: just return false on failure from stop() and let it's caller decide what action to do (exit/throw an exception)

CVS-96324 created for that